File: src/main/java/com/googongill/aditory/controller/dto/user/response/UserInfoResponse.java
package com.googongill.aditory.controller.dto.user.response;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import com.googongill.aditory.domain.User;
import lombok.Builder;

@Builder
@JsonSerialize
@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
public class UserInfoResponse {
    private Long userId;
    private String username;
    private String nickname;
    private Integer aditoryPower;

    public static UserInfoResponse of(User user) {
        Integer totalLikes = user.getCategories().stream()
                .mapToInt(category -> category.getCategoryLikes().size())
                .sum();
        Integer aditoryPower = user.getCategories().size() + user.getLinks().size() + totalLikes;

        return UserInfoResponse.builder()
                .userId(user.getId())
                .username(user.getUsername())
                .nickname(user.getNickname())
                .aditoryPower(aditoryPower)
                .build();
    }
}

File: src/main/java/com/googongill/aditory/controller/dto/user/response/UserInfoResponse.kt
package com.googongill.aditory.controller.dto.user.response

import com.fasterxml.jackson.annotation.JsonAutoDetect
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import com.googongill.aditory.domain.Category
import com.googongill.aditory.domain.User

@JsonSerialize
@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
data class UserInfoResponse(
    val userId: Long?,
    val username: String,
    val nickname: String?,
    val aditoryPower: Int?
) {

    companion object {
        fun of(user: User): UserInfoResponse {
            val totalLikes = user.categories.stream()
                .mapToInt { category: Category -> category.categoryLikes.size }
                .sum()
            val aditoryPower = user.categories.size + user.links.size + totalLikes
            return UserInfoResponse(
                userId = user.id,
                username = user.username,
                nickname = user.nickname,
                aditoryPower = aditoryPower
            )

        }
    }

}